syntax = "proto3";

package foreverbull.service;

option go_package = "github.com/lhjnilsson/foreverbull/pkg/pb/service";

message Algorithm {
    message FunctionParameter {
        string key = 1;
        optional string defaultValue = 2;
        optional string value = 3;
        string valueType = 4;
    }
    message Function {
        string name = 1;
        repeated FunctionParameter parameters = 2;
        bool parallelExecution = 3;
        bool runFirst = 4;
        bool runLast = 5;
    }
    string file_path = 1;
    repeated Function functions = 2;
    repeated string namespaces = 3;
}

message ExecutionConfiguration {
    message FunctionParameter {
        string key = 1;
        string value = 2;
    }
    message Function {
        string name = 1;
        repeated FunctionParameter parameters = 2;
    }
    int32 brokerPort = 1;
    int32 namespacePort = 2;
    string databaseURL = 3;
    repeated Function functions = 4;
}
